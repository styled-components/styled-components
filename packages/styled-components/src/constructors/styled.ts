import createStyledComponent from '../models/StyledComponent';
import { IStyledComponentFactory, WebTarget } from '../types';
import domElements from '../utils/domElements';
import constructWithOptions from './constructWithOptions';

const styled = <Props>(tag: WebTarget) =>
  constructWithOptions<IStyledComponentFactory, Props>(createStyledComponent, tag);

type BaseStyled = typeof styled;

const enhancedStyled = styled as BaseStyled &
  {
    [key in typeof domElements[number]]: ReturnType<BaseStyled>;
  };

// Shorthands for all valid HTML Elements
domElements.forEach(domElement => {
  enhancedStyled[domElement] = styled(domElement);
});

export default enhancedStyled;
