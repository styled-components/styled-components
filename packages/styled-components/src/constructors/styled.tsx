import createStyledComponent from '../models/StyledComponent';
import { WebTarget } from '../types';
import domElements from '../utils/domElements';
import constructWithOptions, { Styled } from './constructWithOptions';

const baseStyled = <Target extends WebTarget>(tag: Target) =>
  constructWithOptions<'web', Target>(createStyledComponent, tag);

const styled = baseStyled as typeof baseStyled & {
  [E in keyof JSX.IntrinsicElements]: Styled<'web', E, JSX.IntrinsicElements[E]>;
};

// Shorthands for all valid HTML Elements
domElements.forEach(domElement => {
  // @ts-expect-error someday they'll handle imperative assignment properly
  styled[domElement] = baseStyled(domElement);
});

export default styled;
