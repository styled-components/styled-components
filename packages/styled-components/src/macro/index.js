// @flow
import { addDefault, addNamed } from '@babel/helper-module-imports';
import traverse from '@babel/traverse';
import { createMacro } from 'babel-plugin-macros';
import babelPlugin from 'babel-plugin-styled-components';
import escapeRegExp from 'lodash/escapeRegExp';

function styledComponentsMacro({
  references,
  state,
  babel: { types: t },
  config: { importModuleName = 'styled-components', ...config } = {},
}) {
  const program = state.file.path;

  // FIRST STEP : replace `styled-components/macro` by `styled-components
  // references looks like this
  // { default: [path, path], css: [path], ... }
  let customImportName;
  Object.keys(references).forEach(refName => {
    // generate new identifier
    let id;
    if (refName === 'default') {
      id = addDefault(program, importModuleName, { nameHint: 'styled' });
      customImportName = id;
    } else {
      id = addNamed(program, refName, importModuleName, { nameHint: refName });
    }

    // update references with the new identifiers
    references[refName].forEach(referencePath => {
      // eslint-disable-next-line no-param-reassign
      referencePath.node.name = id.name;
    });
  });

  // SECOND STEP : apply babel-plugin-styled-components to the file
  const stateWithOpts = {
    ...state,
    opts: {
      ...config,
      topLevelImportPaths: (config.topLevelImportPaths || []).concat(`$${escapeRegExp(importModuleName)}^`),
    },
    customImportName,
  };
  traverse(program.parent, babelPlugin({ types: t }).visitor, undefined, stateWithOpts);
}

export default createMacro(styledComponentsMacro, {
  configName: 'styledComponents',
});
