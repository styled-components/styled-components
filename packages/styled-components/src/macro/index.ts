import { types } from '@babel/core';
import { addDefault, addNamed } from '@babel/helper-module-imports';
import traverse from '@babel/traverse';
import { createMacro, MacroParams } from 'babel-plugin-macros';
import babelPlugin from 'babel-plugin-styled-components';

function styledComponentsMacro({
  references,
  state,
  babel: { types: t },
  config: { importModuleName = 'styled-components', ...config } = {},
}: MacroParams) {
  const program = state.file.path;

  // FIRST STEP : replace `styled-components/macro` by `styled-components
  // references looks like this
  // { default: [path, path], css: [path], ... }
  let customImportName;
  Object.keys(references).forEach(refName => {
    // generate new identifier
    let id: types.Identifier;
    if (refName === 'default') {
      id = addDefault(program, importModuleName, { nameHint: 'styled' });
      customImportName = id;
    } else {
      id = addNamed(program, refName, importModuleName, { nameHint: refName });
    }

    // update references with the new identifiers
    references[refName].forEach(referencePath => {
      // eslint-disable-next-line no-param-reassign
      (referencePath.node as types.Identifier).name = id.name;
    });
  });

  // SECOND STEP : apply babel-plugin-styled-components to the file
  const stateWithOpts = {
    ...state,
    opts: {
      ...config,
      topLevelImportPaths: (config.topLevelImportPaths || []).concat(importModuleName),
    },
    customImportName,
  };
  traverse(program.parent, babelPlugin({ types: t }).visitor, undefined, stateWithOpts);
}

export default createMacro(styledComponentsMacro, {
  configName: 'styledComponents',
});
