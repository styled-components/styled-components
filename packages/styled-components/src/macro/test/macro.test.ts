import babel from '@babel/core';
import plugin from 'babel-plugin-macros';
import pluginTester from 'babel-plugin-tester';

const styledExampleCode = `
import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`
`;

const customStyledExampleCode = `
import myStyled from '../../macro'
import styled from 'another-package'

myStyled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`
`;

const cssExampleCode = `
import { css } from '../../macro'

css\`
  color: \${props => (props.whiteColor ? 'white' : 'black')};
\`
`;

const keyframesExampleCode = `
import { keyframes } from '../../macro'

keyframes\`
  0% { opacity: 0; }
  100% { opacity: 1; }
\`
`;

const createGlobalStyleExampleCode = `
import { createGlobalStyle } from '../../macro'

createGlobalStyle\`
  background: red;
\`
`;

const ThemeProviderExampleCode = `
import { ThemeProvider } from '../../macro'

React.createComponent(
  ThemeProvider,
  { theme: { color: 'red' }},
  'hello'
)
`;

const extendsExampleCode = `
import React from 'react'
import styled from '../../macro'

const Hello = () => React.createComponent(div, null, 'hello')

styled(Hello)\`
  background: red;
\`
`;

const multipleImportsExampleCode = `
import styled, { css } from '../../macro'
`;

const requireExampleCode = `
const styled = require('../../macro')

styled.div\`
  background: red;
\`
`;

const withTypeImportExampleCode = `
import { DefaultTheme } from '../../macro'
`;

const withTypeAndStandardImportExampleCode = `
  import styled, { DefaultTheme } from '../../macro'
`;

const cssPropExampleCode = `
import styled from '../../macro'
import React from 'react';
function Foo() {
  return <div css="color: red;" />;
}
`;

const cssPropOverridingComponentExampleCode = `
import React from 'react';
import styled from '../../macro'
const Thing = styled.div\`
  color: blue;
\`;
function Foo() {
  return <Thing css="color: red;" />;
}
`;

pluginTester({
  title: 'macro',
  plugin,
  snapshot: true,
  babelOptions: {
    babelrc: false,
    filename: __filename,
    presets: [
      '@babel/preset-react',
      ['@babel/preset-typescript', { allExtensions: true, isTSX: true }],
    ],
  },
  babel,
  tests: {
    'should work with styled': {
      code: styledExampleCode,
    },
    'should work with custom import name': {
      code: customStyledExampleCode,
    },
    'should work with { css }': {
      code: cssExampleCode,
    },
    'should work with { keyframes }': {
      code: keyframesExampleCode,
    },
    'should work with { createGlobalStyle }': {
      code: createGlobalStyleExampleCode,
    },
    'should work with { ThemeProvider }': {
      code: ThemeProviderExampleCode,
    },
    'should work when extending a component': {
      code: extendsExampleCode,
    },
    'should work with require() to import styled-components': {
      code: requireExampleCode,
    },
    'should work with multiple imports': {
      code: multipleImportsExampleCode,
    },
    'should work with types': {
      code: withTypeImportExampleCode,
    },
    'should work with types alongside import': {
      code: withTypeAndStandardImportExampleCode,
    },
    'should not add componentId with a config disabling ssr': {
      code: styledExampleCode,
      pluginOptions: {
        styledComponents: {
          ssr: false,
        },
      },
    },
    'should work with the css prop': { code: cssPropExampleCode },
    'should work with the css prop overriding an existing styled-component': {
      code: cssPropOverridingComponentExampleCode,
    },
    'should use a custom import with importModuleName': {
      code: styledExampleCode,
      pluginOptions: {
        styledComponents: {
          importModuleName: '@xstyled/styled-components',
        },
      },
    },
  },
});
