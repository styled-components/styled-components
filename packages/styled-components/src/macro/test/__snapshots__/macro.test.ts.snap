// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macro should not add componentId with a config disabling ssr: should not add componentId with a config disabling ssr 1`] = `

import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from 'styled-components';

_styled.div.withConfig({
  displayName: 'macrotest',
})(['background:', ';'], p => (p.error ? 'red' : 'green'));


`;

exports[`macro should use a custom import with importModuleName: should use a custom import with importModuleName 1`] = `

import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from '@xstyled/styled-components';

_styled.div.withConfig({
  displayName: 'macrotest',
  componentId: 'sc-17fdca0-0',
})(['background:', ';'], p => (p.error ? 'red' : 'green'));


`;

exports[`macro should work when extending a component: should work when extending a component 1`] = `

import React from 'react'
import styled from '../../macro'

const Hello = () => React.createComponent(div, null, 'hello')

styled(Hello)\`
  background: red;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from 'styled-components';
import React from 'react';

const Hello = () => React.createComponent(div, null, 'hello');

_styled(Hello).withConfig({
  displayName: 'macrotest',
  componentId: 'sc-17fdca0-0',
})(['background:red;']);


`;

exports[`macro should work with { ThemeProvider }: should work with { ThemeProvider } 1`] = `

import { ThemeProvider } from '../../macro'

React.createComponent(
  ThemeProvider,
  { theme: { color: 'red' }},
  'hello'
)

      ↓ ↓ ↓ ↓ ↓ ↓

import { ThemeProvider as _ThemeProvider } from 'styled-components';
React.createComponent(
  _ThemeProvider,
  {
    theme: {
      color: 'red',
    },
  },
  'hello'
);


`;

exports[`macro should work with { createGlobalStyle }: should work with { createGlobalStyle } 1`] = `

import { createGlobalStyle } from '../../macro'

createGlobalStyle\`
  background: red;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import { createGlobalStyle as _createGlobalStyle } from 'styled-components';
_createGlobalStyle\`
  background: red;
\`;


`;

exports[`macro should work with { css }: should work with { css } 1`] = `

import { css } from '../../macro'

css\`
  color: \${props => (props.whiteColor ? 'white' : 'black')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import { css as _css } from 'styled-components';

_css(['color:', ';'], props => (props.whiteColor ? 'white' : 'black'));


`;

exports[`macro should work with { keyframes }: should work with { keyframes } 1`] = `

import { keyframes } from '../../macro'

keyframes\`
  0% { opacity: 0; }
  100% { opacity: 1; }
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import { keyframes as _keyframes } from 'styled-components';

_keyframes(['0%{opacity:0;}100%{opacity:1;}']);


`;

exports[`macro should work with custom import name: should work with custom import name 1`] = `

import myStyled from '../../macro'
import styled from 'another-package'

myStyled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from 'styled-components';

_styled.div.withConfig({
  displayName: 'macrotest',
  componentId: 'sc-17fdca0-0',
})(['background:', ';'], p => (p.error ? 'red' : 'green'));


`;

exports[`macro should work with multiple imports: should work with multiple imports 1`] = `

import styled, { css } from '../../macro'

      ↓ ↓ ↓ ↓ ↓ ↓

import { css as _css } from 'styled-components';
import _styled from 'styled-components';


`;

exports[`macro should work with require() to import styled-components: should work with require() to import styled-components 1`] = `

const styled = require('../../macro')

styled.div\`
  background: red;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from 'styled-components';

_styled.div.withConfig({
  displayName: 'macrotest',
  componentId: 'sc-17fdca0-0',
})(['background:red;']);


`;

exports[`macro should work with styled: should work with styled 1`] = `

import styled from '../../macro'

styled.div\`
  background: \${p => (p.error ? 'red' : 'green')};
\`

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled from 'styled-components';

_styled.div.withConfig({
  displayName: 'macrotest',
  componentId: 'sc-17fdca0-0',
})(['background:', ';'], p => (p.error ? 'red' : 'green'));


`;

exports[`macro should work with the css prop overriding an existing styled-component: should work with the css prop overriding an existing styled-component 1`] = `

import React from 'react';
import styled from '../../macro'
const Thing = styled.div\`
  color: blue;
\`;
function Foo() {
  return <Thing css="color: red;" />;
}

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled2 from 'styled-components';
import _styled from 'styled-components';
import React from 'react';

const Thing = _styled.div.withConfig({
  displayName: 'macrotest__Thing',
  componentId: 'sc-17fdca0-0',
})(['color:blue;']);

var _StyledThing = _styled(Thing).withConfig({
  displayName: 'macrotest___StyledThing',
  componentId: 'sc-17fdca0-1',
})(['color:red;']);

function Foo() {
  return /*#__PURE__*/ React.createElement(_StyledThing, null);
}


`;

exports[`macro should work with the css prop: should work with the css prop 1`] = `

import styled from '../../macro'
import React from 'react';
function Foo() {
  return <div css="color: red;" />;
}

      ↓ ↓ ↓ ↓ ↓ ↓

import _styled2 from 'styled-components';
import _styled from 'styled-components';
import React from 'react';

function Foo() {
  return /*#__PURE__*/ React.createElement(_StyledDiv, null);
}

var _StyledDiv = _styled('div').withConfig({
  displayName: 'macrotest___StyledDiv',
  componentId: 'sc-17fdca0-0',
})(['color:red;']);


`;

exports[`macro should work with types alongside import: should work with types alongside import 1`] = `

import styled, { DefaultTheme } from '../../macro'

      ↓ ↓ ↓ ↓ ↓ ↓

import { DefaultTheme as _DefaultTheme } from 'styled-components';
import _styled from 'styled-components';


`;

exports[`macro should work with types: should work with types 1`] = `

import { DefaultTheme } from '../../macro'

      ↓ ↓ ↓ ↓ ↓ ↓

import { DefaultTheme as _DefaultTheme } from 'styled-components';


`;
